### PR TITLE
Add argument to search run to filter on columns names

### PR DESCRIPTION
**Pull request by jdlesage**
## What changes are proposed in this pull request?

New parameter to search run method to filter on columns
Implemented in sql and file store

## How is this patch tested?

(Details)

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 1]
```diff
@@ -853,3 +859,41 @@ def _get_orderby_clauses(order_by_list, session):
...
+    # groupby type of column
+    if columns_to_whitelist is not None:
+        words = [column.split('.') for column in sorted(columns_to_whitelist)]
+        columns = [(word[0], '.'.join(word[1:])) for word in words]
+        for column_type, columns_name in itertools.groupby(columns, key=operator.itemgetter(0)):
+            list_columns_to_whitelist = [word[1] for word in columns_name]
+            queries[column_type] = queries[column_type].filter(
+                objects[column_type].key.in_(list_columns_to_whitelist))
+
+    _logger.info('new query!')
```
Comment (jdlesage) [inline mlflow/store/tracking/sqlalchemy_store.py:886]: Note to myself.... remove the logger

File: mlflow/store/tracking/file_store.py [hunk 2]
```diff
@@ -798,3 +799,16 @@ def record_logged_model(self, run_id, mlflow_model):
 
     def update_artifacts_location(self, run_id, new_artifacts_location):
         raise ValueError("Artifacts cannot be moved with FileStore backend")
+
+
+def _filter_columns(run, columns_to_whitelist):
+    run_data = run.data
+    if columns_to_whitelist is not None:
```
Comment (vguerra) [inline mlflow/store/tracking/file_store.py:806]: nitpick: what about inverting the logic of this condition so that you reduce the number of lines of nested code? wdyt?
```suggestion
    if columns_to_whitelist is None:
        return run
    ....
```

Comment (jdlesage) [reply mlflow/store/tracking/file_store.py]: Done

File: mlflow/tracking/client.py [hunk 3]
```diff
@@ -331,7 +331,8 @@ def restore_run(self, run_id):
         self._tracking_client.restore_run(run_id)
 
     def search_runs(self, experiment_ids, filter_string="", run_view_type=ViewType.ACTIVE_ONLY,
-                    max_results=SEARCH_MAX_RESULTS_DEFAULT, order_by=None, page_token=None):
+                    max_results=SEARCH_MAX_RESULTS_DEFAULT, order_by=None, page_token=None,
+                    columns_to_whitelist=None):
```
Comment (t-henri) [inline mlflow/tracking/client.py:335]: It would be nice to add tests for this

File: mlflow/tracking/fluent.py [hunk 4]
```diff
@@ -409,7 +409,8 @@ def get_artifact_uri(artifact_path=None):
 
 
 def search_runs(experiment_ids=None, filter_string="", run_view_type=ViewType.ACTIVE_ONLY,
-                max_results=SEARCH_MAX_RESULTS_PANDAS, order_by=None):
+                max_results=SEARCH_MAX_RESULTS_PANDAS, order_by=None,
+                columns_to_whitelist=None):
```
Comment (t-henri) [inline mlflow/tracking/fluent.py:413]: It would be nice to add tests for this

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 5]
```diff
@@ -372,9 +376,9 @@ def _get_eager_run_query_options():
             # of the eager loading procedure. For more information about relationship loading
             # techniques, see https://docs.sqlalchemy.org/en/13/orm/
             # loading_relationships.html#relationship-loading-techniques
-            sqlalchemy.orm.subqueryload(SqlRun.latest_metrics),
-            sqlalchemy.orm.subqueryload(SqlRun.params),
-            sqlalchemy.orm.subqueryload(SqlRun.tags),
+            sqlalchemy.orm.selectinload(SqlRun.latest_metrics),
+            sqlalchemy.orm.selectinload(SqlRun.params),
+            sqlalchemy.orm.selectinload(SqlRun.tags),
```
Comment (t-henri) [inline mlflow/store/tracking/sqlalchemy_store.py:381]: Why did you change these if we don't use the function anymore ?

**Comment by rom1504** [comment]
I think it would make sense to do a benchmark and add it in the description of that PR to motivate adding that feature. (especially when submitting this PR to databricks)
The results were pretty good when I did that a while ago.

**Comment by jdlesage** [comment]
@rom1504 I agree. On my workstation with your project the results are encouraging:

100 runs => 3.67 s all columns . 120 ms no columns. ~30ms per column
it seems linear : time is  O(columns * runs)

There is a hard limit to 50k run in rest_store
I tried 2 columns on this 50k runs. 1min to fetch them  all

Our prod servers are 30% faster than my workstation (from my measures)

**Comment by rom1504** [comment]
That seems pretty good indeed. We usually need maximum 10 columns. That's
the maximum that makes sense visually in the ui, and in our autovalidator
we need about the same number, and it can be about the same for correlation
analysis between metrics.
Thank you for working on this!

File: mlflow/tracking/fluent.py (hunk 4 above)
Comment (jdlesage) [reply mlflow/tracking/fluent.py:413]: There are no logic in the functions (just forward the argument to client through _get_paginated_runs). Do you agree to not test that ?

**Comment by jdlesage** [comment]
Here some first benchmark: The experiment has 1500 columns. 100 runs search takes 3.67s

**Comment by jdlesage** [comment]
@rom1504 I will merge it. The last commit makes the query 2 times faster with the number of rows 
image

**Comment by rom1504** [comment]
Nice !
Are you not worried to have conflicts if you don't also merge it upstream ?
